In [1]:
import pandas as pd
import numpy as np

from sklearn.model_selection import KFold, cross_validate
from sklearn.linear_model import LinearRegression

In [2]:
df = pd.read_csv("../Airlinedataset.csv")

features = [
    "days_to_departure",
    "seats_remaining",
    "historical_demand",
    "competitor_price",
    "booking_velocity",
    "is_weekend",
    "flight_capacity"
]

target = "ticket_price"

X = df[features]
y = df[target]

In [3]:
kf = KFold(
    n_splits=5,
    shuffle=True,
    random_state=42
)

model = LinearRegression()

In [4]:
cv_results = cross_validate(
    model,
    X,
    y,
    cv=kf,
    scoring={
        "mae": "neg_mean_absolute_error",
        "rmse": "neg_root_mean_squared_error",
        "r2": "r2"
    },
    return_train_score=True
)

In [5]:
validation_results = pd.DataFrame({
    "Fold": range(1, 6),
    "MAE": -cv_results["test_mae"],
    "RMSE": -cv_results["test_rmse"],
    "R2": cv_results["test_r2"]
})

validation_results

,Fold,MAE,RMSE,R2
0,1,228.291103,278.180900,0.958463
1,2,218.146243,275.419866,0.952737
2,3,235.089687,291.090978,0.953809
3,4,232.928963,282.641411,0.955050
4,5,217.605204,275.631079,0.950406


In [6]:
summary = pd.DataFrame({
    "Metric": ["MAE", "RMSE", "R2"],
    "Mean": [
        validation_results["MAE"].mean(),
        validation_results["RMSE"].mean(),
        validation_results["R2"].mean()
    ],
    "Std": [
        validation_results["MAE"].std(),
        validation_results["RMSE"].std(),
        validation_results["R2"].std()
    ]
})

summary

,Metric,Mean,Std
0,MAE,226.412240,8.17292
1,RMSE,280.592847,6.54988
2,R2,0.954093,0.00298


In [7]:
print("5-Fold Cross-Validation Results")
print("--------------------------------")

print(
    f"MAE:  "
    f"{validation_results['MAE'].mean():.2f} "
    f"+/- "
    f"{validation_results['MAE'].std():.2f}"
)

print(
    f"RMSE: "
    f"{validation_results['RMSE'].mean():.2f} "
    f"+/- "
    f"{validation_results['RMSE'].std():.2f}"
)

print(
    f"R²:   "
    f"{validation_results['R2'].mean():.4f} "
    f"+/- "
    f"{validation_results['R2'].std():.4f}"
)

5-Fold Cross-Validation Results
--------------------------------
MAE:  226.41 +/- 8.17
RMSE: 280.59 +/- 6.55
R²:   0.9541 +/- 0.0030


## Cross-Validation Findings

Five-fold cross-validation was used to evaluate the stability of the Linear Regression model.

| Metric | Mean | Standard Deviation |
|---|---:|---:|
| MAE | 226.41 | 8.17 |
| RMSE | 280.59 | 6.55 |
| R² | 0.9541 | 0.0030 |

The cross-validation results are close to the original 80/20 holdout results:

- Holdout MAE: 228.29
- Cross-validation MAE: 226.41 ± 8.17

- Holdout RMSE: 278.18
- Cross-validation RMSE: 280.59 ± 6.55

- Holdout R²: 0.9585
- Cross-validation R²: 0.9541 ± 0.0030

The relatively small variation across folds indicates that the model's predictive performance is reasonably stable on different subsets of this dataset.

Cross-validation does not establish real-world temporal validity because the dataset does not contain chronological booking observations.